packages/store-ui/src/molecules/Modal/Modal.tsx

```diff
@@ -0,0 +1,140 @@
+/* eslint-disable jsx-a11y/no-noninteractive-tabindex */
+/*
+ * Disable the eslint rule to use sentinel html elements that aren't interactive.
+ */
+import type {
+  AriaAttributes,
+  HTMLAttributes,
+  KeyboardEvent,
+  MouseEvent,
+  PropsWithChildren,
+} from 'react'
+import React, { useRef } from 'react'
+import { createPortal } from 'react-dom'
+
+import Overlay from '../../atoms/Overlay'
+import useTrapFocus from './useTrapFocus'
+
+const composeEventHandler = <EventType extends React.SyntheticEvent | Event>(
+  handler: ((event: EventType) => void) | undefined,
+  internalHandler: (event: EventType) => void
+) => {
+  return (event: EventType) => {
+    handler?.(event)
+    if (!event.defaultPrevented) {
+      internalHandler(event)
+    }
+  }
+}
+
+interface ModalContentProps extends HTMLAttributes<HTMLDivElement> {
+  testId?: string
+  onDismiss?: (event: MouseEvent | KeyboardEvent) => void
+}
+
+const ModalContent = ({ children, ...props }: ModalContentProps) => {
+  const trapFocusRef = useRef<HTMLDivElement>(null)
+  const sentinelStartRef = useRef<HTMLDivElement>(null)
+  const sentinelEndRef = useRef<HTMLDivElement>(null)
+
+  useTrapFocus({
+    sentinelStartRef,
+    sentinelEndRef,
+    trapFocusRef,
+  })
+
+  return (
+    <>
+      <div tabIndex={0} data-testid="sentinelStart" ref={sentinelStartRef} />
+      {/*
+       * This next line is disabled due to the onClick prop.
+       * Even though div isn't clickable, the onClick is required to prevent the event bubbles
+       * until the overlay that calls onDismiss inside the onClick handler.
```

Comment:
This phrase left me a little confused, what did you mean by that?
Is it like:
```suggestion
       * until the overlay, which calls onDismiss inside the onClick handler, is clicked.
```
?